# Finding numbers that are simultaneously squares and other polygons

In [1]:
from __future__ import division

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math as mt

/Users/marcoscscarreira/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')


### Formula for the nth polygonal number (where sd is the number of sides):

In [3]:
def polnr(n, sd):
    return int(n*((sd-2)*n-(sd-4))/2)

Triangular numbers:

In [4]:
[polnr(n,3) for n in range(11)]

[0, 1, 3, 6, 10, 15, 21, 28, 36, 45, 55]

Square numbers:

In [5]:
[polnr(n,4) for n in range(11)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

Pentagonal numbers:

In [6]:
[polnr(n,5) for n in range(11)]

[0, 1, 5, 12, 22, 35, 51, 70, 92, 117, 145]

Hexagonal numbers:

In [7]:
[polnr(n,6) for n in range(10)]

[0, 1, 6, 15, 28, 45, 66, 91, 120, 153]

### Which numbers are simultaneously triangular and square?

In [8]:
import pyprimes.factors as pypr

In [9]:
triset=set([polnr(n,3) for n in range(1,1000001)])

In [10]:
squset=set([polnr(n,4) for n in range(1,1000001)])

In [19]:
trisquset=list(set.intersection(triset,squset))

In [21]:
trisquset.sort()

These are the numbers that are simultaneously triangular and square:

In [22]:
trisquset

[1, 36, 1225, 41616, 1413721, 48024900, 1631432881, 55420693056]

These are the square roots of the numbers that are simultaneously triangular and square:

In [30]:
sqrn=map(lambda x: int(mt.sqrt(x)),trisquset)

In [31]:
sqrn

[1, 6, 35, 204, 1189, 6930, 40391, 235416]

These are the prime factors of the square roots of the numbers that are simultaneously triangular and square:

In [33]:
map(pypr.factorise,sqrn)

[[1],
 [2, 3],
 [5, 7],
 [2, 2, 3, 17],
 [29, 41],
 [2, 3, 3, 5, 7, 11],
 [13, 13, 239],
 [2, 2, 2, 3, 17, 577]]

Let's look at the odd numbers:

In [41]:
sqrno=sqrn[::2]

In [42]:
sqrno

[1, 35, 1189, 40391]

35 is factored as 5x7 = (6-1)x(6+1)
1189 is factored as 29x41 = (35-6)x(35+6)
40391 is factored as 169x239 = (204-35)x(204+35)

Let's look at the even numbers divided by 6:

In [39]:
sqrne6=map(lambda x: int(x/6),sqrn[1::2])

In [40]:
sqrne6

[1, 34, 1155, 39236]

Now the difference between the odd numbers and the even numbers divided by 6 is the shifted list of even numbers divided by 6

In [46]:
[sqrno[j]-sqrne6[j] for j in range(4)]

[0, 1, 34, 1155]

One can use these results and find: 

In [56]:
def nextsqroot(x):
    if x==1:
        return 1
    else:
        return 3*nextsqroot(x-1)+int(mt.sqrt((3**2-1)*(nextsqroot(x-1)**2)+1))

In [59]:
[nextsqroot(j) for j in range(1,9)]

[1, 6, 35, 204, 1189, 6930, 40391, 235416]

In [60]:
sqrn

[1, 6, 35, 204, 1189, 6930, 40391, 235416]